README.md
# GAME-ACTION-DaveGnukem
Dave Gnukem is a 2D scrolling platform shooter. It is inspired by and similar to Duke Nukem 1. 

LICENSE
===============
* GNU LGPL

COMPILE TOOLS
===============
* Req: Scitech SNAP - DART - Libc
 
AUTHORS
===============
* Ported by Andreas

LINKS
===============
* https://www.os2world.com/games/index.php/native-games/action-2/42-dave-gnukem


HISTORY.txt
2002/09/25 - Version 0.54e
 - Scrapped 'lives'
 - Some more tileset additions/changes from Evil Mr Henry
 - New cheat keys in game (god mode).

2002/09/24 - Version 0.54d
 - Added some more new sprites from Evil Mr Henry
 - Fixed level-editor crash-when-view-at-bottom-of-map bug

2002/09/21 - Version 0.54c
 - Begin integrating integrated level editor from Vytautas Shaltenis (rtfb).
 - Added some new sprites from Evil Mr Henry

2002/09/15
 - Fixed CR/LF text file reading problem in Linux
 - Fixed a problem with the TGA loader

2002/09/09 - Version 0.54b
 - Added levels from "TOG of Spookware"

2001/09/23 - Version 0.54
 - Dynamite explosion sound

2001/09/22
 - Monster bullets

2001/09/20
 - Get it compiling on Linux again
 - More load/save game stuff

2001/09/19
 - Integrate SDL_mixer sound support contributions from Kent Mein

2001/09/03
 - Made keys redefinable

2001/09/02
 - Misc changes and bugfixes
 - Ability to enter high score username
 - Added stationary spiked ball
 - Added basic framework for redefining keys
 - Added configuration file stuff
 - Added some very rudimentary save/load game stuff
 - "Dust" effect as hero lands

2001/08/11
 - Misc changes (some BeOS changes)
 - Flame monster thing (similar to dumbbot)

2001/07/25
 - Cleaned up (genericized) teleport and level-exit code
 - Exit now animates

2001/07/24 - Version 0.53

2001/07/22
 - Ported to SDL (still wrapped though)
 - Minor editor improvements

2001/07/15 - Version 0.52

2001/07/14
 - Misc small changes/fixes/graphics improvements etc

2001/07/13
 - Small changes and fixes
 - Shootable coke can
 - Proper "firepower"

2001/07/12
 - Fans

2001/07/11
 - Rewrote (cleaned up) bullet code
 - Misc bugfixes

2001/07/10
 - Falling block disappears when hits floor
 - Bullet stops when hits solid stuff
 - Got rid of djArray

2001/07/09
 - Dynamite (flames in gray boxes)

2001/07/06
 - Rewrote conveyor belt code
 - Side flame throw blocks (flames look a bit like condoms though :/)
 - Level background image

2001/07/04
 - Rewrote (cleaned up) monster (robot) code
 - Rewrote (cleaned up) lift code
 - Misc cleanups (e.g. palette stuff killed for good)

2001/07/03 - Version 0.51c
 - Reasonable cleanup: Replaced stupid "objects" array stuff
   with CPickup class, also powerboots etc. Added "replace"
   utility for level chars. Minor graphics improvements.

2001/07/01
 - Floating balloon
 - Acme falling block
 - Disk
 - Slightly improved teleporter appearance

2001/06/21 - Version 0.51b
 - Spikes
 - "Example" level

2001/06/20
 - Bonus when collecting G,N,U,K,E,M in order
 - Bonus when shooting all cameras in a level
 - New collectables (e.g. rugby ball, joystick, radio etc)
 - Misc graphics improvements

2001/06/17 - Version 0.51
 - Release 0.51
   Added a "thing factory" to make it a bit easier to add new thing types.
   Completely rewrote door/key code. Can now pick up letters. Added shootable
   solid blocks. Can now shoot security cameras. Added green crawly things.
   Added purple flags (also in boxes). Added shootable banana. Added high
   score list (almost)

2001/06/16
 - Got rid of djGameLib :) Cleaned up quite a bit. Added a bouncing spiky ball.

2001/06/14
 - Some more misc cleanups. Put new version on web page.

2001/06/03
 - Misc cleanups, misc graphics improvements, added doxygen support

2000/01/03 - Version 0.5
 - Release 0.5

1999/12/30
 - Nice crates :)
 - "Macros" for the level editor

1999/12/29
 - More cleaning up (surprise)
 - Toggle background blocks view on level editor
 - Some graphics improvements
 - 16-bit color mode support

1999/12/28
 - Added some more sound effects
 - Moved the input stuff into the DJG library
 - A whole lot more cleaning up

1999/12/27
 - Ported level editor to also work on Windows
 - Added "lives"
 - Added basic sound support to Windows version

1999/12/26
 - Win32 version :)
 - Added an instructions screen
 - More code refactoring and cleaning up

1999/12/25
 - More refactoring and cleaning up. Graph.h/cpp now platform-independent

1999/12/24
 - Extensive refactoring to clean up and make crossplatform
 - Started adding support for 32-bit images, fonts, sprites etc
 - Sprites can now be in TGA format

1999/06/07
 - More structural changes to make cross-platform

1999/05/29
 - Start to port game to "djg", cross-platform library

1999/03/31
 - Boxes now fall

1999/03/23 - Version 0.44c
 - Added another level from Steve Merrifield
 - Added boxes
 - Started adding infrastructure for bonus letters

1999/03/22 - Version 0.44b
 - Added level from Steve Merrifield

1999/03/08 - Version 0.44
 - Fixed editor bug of not displaying sprite type properly
 - Added ability to select levelset in editor

1999/03/07
 - Updated Felix Richters english documentation and replaced README
   with it.

1999/02/28 - Version 0.43
 - Ability to select levelset (mission) from menu.
 - New levelset infrastructure virtually in now.

1999/02/27
 - Yet more sprite code. *sigh* .. Almost done. RLE is dead.
   Long live bitmask. For now, anyway.

1999/02/26
 - Aaargh! Still more spriteset code.

1999/02/25
 - More spriteset code.

1999/02/24
 - More spriteset code.

1999/02/21
 - More level set stuff. Re-writing sprite code.

1999/02/18
 - Some more work on level set stuff.

1999/02/17
 - Started adding "level set" infrastructure. Source is bloating :/

1999/02/16
 - Fixed 24-bit mode with the editor

1999/02/14
 - Improved tracing/message system (log.h/log.cpp)
 - Slight speed optimizing: RLE algorithm for sprite transparency

1999/02/07
 - Added support for 16-bit color modes; fixed 24-bit support

1999/01/31
 - Added "double-size" window mode
 - Added powerboots :)
 - Added friendlier "flags edit" feature to sprite editor

1999/01/30
 - Created "template" example level

block.h
/*!
\file    block.h
\brief   Level block types
\author  David Joffe

Copyright (C) 2000-2001 David Joffe

License: GNU GPL Version 2 (*not* "later versions")
*/
#ifndef _BLOCK_H_
#define _BLOCK_H_

//! Block type enumeration
enum EBlockType
{
	TYPE_NOTHING = 0,
	TYPE_1,
	TYPE_2,
	//! 03 Solid, but can come up from underneath
	TYPE_SOFTSOLID,
	TYPE_4,
	TYPE_5,
	//! 06 Dust
	TYPE_DUST,
	//! 07 Lift
	TYPE_LIFT,
	//! 08 Teleporter
	TYPE_TELEPORTER,
	TYPE_09,
	TYPE_10,
	//! 11 Firepower
	TYPE_FIREPOWER,
	TYPE_12,
	//! 13 Conveyor belt
	TYPE_CONVEYOR,
	//! 14 Letter (G,N,U,K,E,M)
	TYPE_LETTER,
	//! 15 Box that opens when shot
	TYPE_BOX,
	TYPE_16,
	//! 17 Generic pickup (food, points etc)
	TYPE_PICKUP,
	TYPE_18,
	//! 19 Power boots
	TYPE_POWERBOOTS,
	//! 20 Coke can
	TYPE_COKE,
	TYPE_21,
	TYPE_22,
	//! 23 Dynamite
	TYPE_DYNAMITE,
	//! 24 Flame throw thingy
	TYPE_FLAMETHROW,
	//! 25 Key. Extra[0]==ID
	TYPE_KEY,
	//! 26 Door. Extra[0]==ID
	TYPE_DOOR,
	//! 27 Door activator. Extra[0]==ID
	TYPE_DOORACTIVATOR,
	//! 28 Fan
	TYPE_FAN,
	//! 29 Access card
	TYPE_ACCESSCARD,
	TYPE_30,
	//! 31 Level exit
	TYPE_EXIT,
	//! 32 Hero starting position
	TYPE_HEROSTART,
	TYPE_33,
	TYPE_34,
	//! 35 Acme falling block
	TYPE_ACME,
	//! 36 Balloon
	TYPE_BALLOON,
	//! 37 Spike
	TYPE_SPIKE,
	//! 38 Monster - Green crawly thing
	TYPE_CRAWLER,
	//! 39 Banana
	TYPE_BANANA,
	//! 40 Security camera
	TYPE_CAMERA,
	//! 41 Solid block which disappears when gets shot
	TYPE_SOFTBLOCK,
	//! 42 Debug block, used for testing stuff out
	TYPE_TEST,
	//! 43 Bouncing spiked ball
	TYPE_SPIKEBALL,
	//! 44 Monster - Dumb robot
	TYPE_ROBOT,
	//! Highest used block value
	TYPE_LASTONE
};

//! Friendly strings for block types \ref EBlockType, used by editor.
extern char * block_type_names[TYPE_LASTONE+1];

#endif
